### ISSUE TITLE
glmer error when using with lapply()

### ISSUE THREAD

**Issue by atyre2** [lme4/lme4#369]
The error arises when running glmer inside lapply().

    library(lme4)

    ## Loading required package: Matrix

    cbpp$obs <- 1:nrow(cbpp)
    (gm2 <- glmer(cbind(incidence, size - incidence) ~ period +
        (1 | herd) +  (1|obs),
                  family = binomial, data = cbpp))

    ## Generalized linear mixed model fit by maximum likelihood (Laplace
    ##   Approximation) [glmerMod]
    ##  Family: binomial  ( logit )
    ## Formula: cbind(incidence, size - incidence) ~ period + (1 | herd) + (1 |  
    ##     obs)
    ##    Data: cbpp
    ##      AIC      BIC   logLik deviance df.resid 
    ## 186.6383 198.7904 -87.3192 174.6383       50 
    ## Random effects:
    ##  Groups Name        Std.Dev.
    ##  obs    (Intercept) 0.8911  
    ##  herd   (Intercept) 0.1840  
    ## Number of obs: 56, groups:  obs, 56; herd, 15
    ## Fixed Effects:
    ## (Intercept)      period2      period3      period4  
    ##      -1.500       -1.226       -1.329       -1.866

    listOfFormulas <- list(
      cbind(incidence, size - incidence) ~ 1 +
        (1 | herd) +  (1|obs),
      cbind(incidence, size - incidence) ~ period +
        (1 | herd) +  (1|obs)
    )
    lapply(listOfFormulas,glmer,family=binomial,data=cbpp)

    ## Error: 'data' not found, and some variables missing from formula environment

    sessionInfo()

    ## R version 3.2.3 (2015-12-10)
    ## Platform: x86_64-w64-mingw32/x64 (64-bit)
    ## Running under: Windows 7 x64 (build 7601) Service Pack 1
    ## 
    ## locale:
    ## [1] LC_COLLATE=English_United States.1252 
    ## [2] LC_CTYPE=English_United States.1252   
    ## [3] LC_MONETARY=English_United States.1252
    ## [4] LC_NUMERIC=C                          
    ## [5] LC_TIME=English_United States.1252    
    ## 
    ## attached base packages:
    ## [1] stats     graphics  grDevices utils     datasets  methods   base     
    ## 
    ## other attached packages:
    ## [1] lme4_1.1-11  Matrix_1.2-3
    ## 
    ## loaded via a namespace (and not attached):
    ##  [1] Rcpp_0.12.2     lattice_0.20-33 digest_0.6.8    MASS_7.3-45    
    ##  [5] grid_3.2.3      nlme_3.1-125    formatR_1.2.1   magrittr_1.5   
    ##  [9] evaluate_0.8    stringi_1.0-1   minqa_1.2.4     nloptr_1.0.4   
    ## [13] rmarkdown_0.9.2 splines_3.2.3   tools_3.2.3     stringr_1.0.0  
    ## [17] yaml_2.1.13     htmltools_0.3   knitr_1.11

The error did not occur when using `lme4_1.1-10`. Creating an anonymous
function works.

    lapply(listOfFormulas,
           function(mm) glmer(mm, ,family=binomial,data=cbpp))

    ## [[1]]
    ## Generalized linear mixed model fit by maximum likelihood (Laplace
    ##   Approximation) [glmerMod]
    ##  Family: binomial  ( logit )
    ## Formula: cbind(incidence, size - incidence) ~ 1 + (1 | herd) + (1 | obs)
    ##    Data: cbpp
    ##      AIC      BIC   logLik deviance df.resid 
    ## 194.4654 200.5414 -94.2327 188.4654       53 
    ## Random effects:
    ##  Groups Name        Std.Dev. 
    ##  obs    (Intercept) 1.160e+00
    ##  herd   (Intercept) 2.905e-05
    ## Number of obs: 56, groups:  obs, 56; herd, 15
    ## Fixed Effects:
    ## (Intercept)  
    ##      -2.526  
    ## 
    ## [[2]]
    ## Generalized linear mixed model fit by maximum likelihood (Laplace
    ##   Approximation) [glmerMod]
    ##  Family: binomial  ( logit )
    ## Formula: cbind(incidence, size - incidence) ~ period + (1 | herd) + (1 |  
    ##     obs)
    ##    Data: cbpp
    ##      AIC      BIC   logLik deviance df.resid 
    ## 186.6383 198.7904 -87.3192 174.6383       50 
    ## Random effects:
    ##  Groups Name        Std.Dev.
    ##  obs    (Intercept) 0.8911  
    ##  herd   (Intercept) 0.1840  
    ## Number of obs: 56, groups:  obs, 56; herd, 15
    ## Fixed Effects:
    ## (Intercept)      period2      period3      period4  
    ##      -1.500       -1.226       -1.329       -1.866

**Comment by dmbates** [comment]
You have a working work-around with the anonymous function so I would stay with that.

tl;dr
I suspect that the failure when calling directly to `glmer` is related to the environment in which the call to `glmer` is evaluated.  By creating an anonymous function you also create an evaluation environment so data argument has a value in the calling frame.

The other failure is a result of what Thomas Lumley has called "the standard non-standard evaluation scheme" used in most calls to the `model.frame` function.  The call to `model.frame` is evaluated not in the frame of the calling function but in its parent frame.  If this seems horribly complicated, it is.  It's one of the prices you pay for lazy evaluation.

**Comment by bbolker** [comment]
Sigh.  I have tracked this down to fcde0d14a4503957a6fc0f3bee4166332ae9b119

```
commit fcde0d14a4503957a6fc0f3bee4166332ae9b119
Author: bbolker <bbolker@gmail.com>
Date:   Sat Nov 14 19:06:42 2015 -0500

    better missing-data error; lots of glmer.nb updating fixes
```

which was my attempt to do a better job detecting when data couldn't be found.  I will see if I can figure out what's going wrong here, as it's obviously a false positive

**Comment by bbolker** [comment]
I have a way to fix this, but it's ugly.  I'll probably push it to a branch.  The better solution (sigh) would be to dig in and diagnose/clean up the problem at a deeper level, looking at all the tests we have so far and adding this case.

**Comment by atyre2** [comment]
Thank you for looking into it. For my purposes the workaround is fine. I posted it here hopefully to save someone else an hour or so of head scratching! `glmer()` [isn't the only function I've run into that has this sort of issue with `lapply()` in the context of comparing models in lists](http://stackoverflow.com/questions/27931317/using-stargazer-with-a-list-of-lm-objects-created-by-lapply-ing-over-a-split-dat) , so I think I'll start teaching students to use anonymous functions all the time.

**Comment by bbolker** [comment]
fixed by e1f21ef

**Comment by mdwhitby** [comment]
I see a fix was created, was it included in version 1.1-12? if so, I still receive the OP's error. Thanks.

**Comment by bbolker** [comment]
No.  1.1-12 is dated 2016-04-16, the fix was 5 months later (September  2016). I do need to push a new CRAN release fairly soon for small/trivial CRAN-cleanup reason.  You could (1) use the workaround described above, (2) install from Github, (3) ask us to push new binary versions to r-forge repo ...

**Comment by geza-kiss** [comment]
Note that you get the same message as the one described by atyre2 when using an lme4 model with MuMIn::pdredge() with a SOCK cluster:

```
> library(lme4)
> library(parallel)
> cluster = makeCluster(3, type = "SOCK")
> full_model = lme4::lmer(... ~ ... + (1 | ...), data = dt, REML = FALSE, na.action = "na.fail")
> MuMIn::pdredge(full_model, cluster)
Fixed term is "(Intercept)"
Warning messages:
1: In lme4::lmer(formula = ... ~ ...  + (1 | study_id),  ... :
'data' not found, and some variables missing from formula environment (model 0 skipped)
...
Error in MuMIn::pdredge(full_model, cluster) : the result is empty
```

(Interestingly, you don't get an error if you use it with MuMIn::dredge(), which is very similar, only not parallelized.)

I installed lme4 from source from your github repo, but I get the same result with that, unfortunately. I do not know what the workaround would be. It would be great if you could address this. Of course the problem may be on the side of MuMIn, but based on the above, I thought I would write to you first.

Thanks!

**Comment by bbolker** [comment]
To @geza-kiss (several years later!)  The solution here is that you need to `clusterExport(cluster, "dt")` in order for this to work; it's explained in the first paragraph of the *Details* section of `?MuMIn::pdredge`.
